eval/eval.go

```diff
@@ -0,0 +1,290 @@
+package main
+
+// This program evaluates rolling hash algorithms on a few different critera:
+//   - It measures the time required to consume a megabyte of random data,
+//     "rolling" and computing a new digest on each byte.
+//   - It measures the likelihood of each output bit being zero,
+//     reporting when the likelihood departs from the expected value of 50%.
+//   - It measures the likelihood that each pair of output bits is correlated,
+//     reporting when the likelihood departs from the expected value of 50%.
+//   - It performs a "strict avalanche" test,
+//     in which varying the input by a single bit should change each output bit with likelihood 50%.
+//
+// Run with -all to test all rolling-hash algorithms.
+// Run with algorithm names to test only those algorithms.
+//
+// To add new algorithms, update the `hashes` map below.
+// It maps an algorithm name to a factory function that produces a ready-to-use `roller`.
+//
+// Note that the abstract `roller` interface adds a few nanoseconds to each method call.
+// This makes the timing results merely an adequate starting point for comparing algorithms.
+
+import (
+	"flag"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/chmduquesne/rollinghash/adler32"
+	"github.com/chmduquesne/rollinghash/bozo32"
+	"github.com/chmduquesne/rollinghash/buzhash32"
+	"github.com/chmduquesne/rollinghash/buzhash64"
+	"github.com/chmduquesne/rollinghash/rabinkarp64"
+	"go4.org/rollsum"
+)
+
+type roller interface {
+	Roll(byte)
+	Digest() uint32
+}
+
+func main() {
+	var (
+		all  = flag.Bool("all", false, "evaluate all hashes")
+		seed = flag.Int64("seed", time.Now().Unix(), "RNG seed")
+	)
+	flag.Parse()
+
+	fmt.Printf("Using RNG seed %d\n", *seed)
+
+	if *all {
+		for name, factory := range hashes {
+			fmt.Printf("%s:\n", name)
+			eval(factory, *seed)
+		}
+	} else {
+		for _, name := range flag.Args() {
+			fmt.Printf("%s:\n", name)
+			eval(hashes[name], *seed)
+		}
+	}
+}
+
+func eval(factory func() roller, seed int64) {
+	src := rand.NewSource(seed)
+	rng := rand.New(src)
+	var megabyte [1024 * 1024]byte
+	rng.Read(megabyte[:])
+
+	var (
+		zeroes       [32]int
+		correlations [32 * 32]int
+	)
+
+	r := factory()
+	var elapsed time.Duration
+	checkpoint := time.Now()
+	for i := 0; i < len(megabyte); i++ {
+		r.Roll(megabyte[i])
+		digest := r.Digest()
+		elapsed += time.Since(checkpoint) // stop timer
```

Comment:
We should probably just measure the running time in a separate pass from the statistical bookkeeping entirely; my sense is that time.Now() on most platforms is probably noticably *more* expensive than the checksum operations, as it requires a system call (though Linux may be an exception here). So I don't trust this not to dilute the differences between hash functions. The results I'm seeing running on my laptop argue for performance not being that different across hashes, but I have a sinking feeling that we may just be measuring time.Now().

Reply:
Quite right, thanks! Fixed.